k8s: Pipes/controls don't work most of the time

See video: http://recordit.co/8BTwXcjnRO

It may (or may not) be k8s-specific since it works in my local k8s cluster.

Apart from the unexpected EOF error I also got a `Bad request` in pipes and a few `Probe _ is not connected right now...`

<img width="1772" alt="screen shot 2016-02-25 at 17 33 22" src="https://cloud.githubusercontent.com/assets/2362916/13328380/351dc6c4-dbe6-11e5-91b1-d68890df9699.png">
<img width="1770" alt="screen shot 2016-02-25 at 17 35 28" src="https://cloud.githubusercontent.com/assets/2362916/13328383/396fa440-dbe6-11e5-819a-30b0167005f5.png">

I repeated the same process while looking at the probe logs and I found this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5075bc]

goroutine 5580 [running]:
github.com/weaveworks/scope/probe/docker.(*registry).execContainer(0xc82062a980, 0xc8203b1cc0, 0x40, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:120 +0x36c
github.com/weaveworks/scope/probe/docker.(*registry).(github.com/weaveworks/scope/probe/docker.execContainer)-fm(0xc8203b1cc0, 0x40, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:160 +0xca
github.com/weaveworks/scope/probe/docker.captureContainerID.func1(0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:149 +0x26d
github.com/weaveworks/scope/probe/controls.HandleControlRequest(0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/controls/controls.go:23 +0x28a
github.com/weaveworks/scope/common/xfer.ControlHandlerFunc.Handle(0x197c9c8, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0xc820952080, 0x0, 0x0)
        /go/src/github.com/weaveworks/scope/common/xfer/controls.go:48 +0x8b
reflect.Value.call(0x140f5c0, 0x14e2d00, 0x13, 0x16b8628, 0x4, 0xc820647ee8, 0x3, 0x3, 0x0, 0x0, ...)
        /usr/local/go/src/reflect/value.go:432 +0x120a
reflect.Value.Call(0x140f5c0, 0x14e2d00, 0x13, 0xc820647ee8, 0x3, 0x3, 0x0, 0x0, 0x0)
        /usr/local/go/src/reflect/value.go:300 +0xb1
net/rpc.(*service).call(0xc820298100, 0xc8202980c0, 0xc8202bc090, 0xc82000a580, 0xc8202ba140, 0x14e6c20, 0xc821324120, 0xd9, 0x11905a0, 0xc820952080, ...)
        /usr/local/go/src/net/rpc/server.go:383 +0x1c1
created by net/rpc.(*Server).ServeCodec
        /usr/local/go/src/net/rpc/server.go:477 +0x4ac
```

--- a/probe/docker/controls.go
+++ b/probe/docker/controls.go
@@ -57,7 +57,7 @@ func (r *registry) attachContainer(containerID string, req xfer.Request) xfer.Re
 	hasTTY := c.HasTTY()
 	id, pipe, err := controls.NewPipe(r.pipes, req.AppID)
 	if err != nil {
-		xfer.ResponseError(err)
+		return xfer.ResponseError(err)
 	}
 	local, _ := pipe.Ends()
 	cw, err := r.client.AttachToContainerNonBlocking(docker_client.AttachToContainerOptions{
@@ -103,12 +103,12 @@ func (r *registry) execContainer(containerID string, req xfer.Request) xfer.Resp
 		Container:    containerID,
 	})
 	if err != nil {
-		xfer.ResponseError(err)
+		return xfer.ResponseError(err)
 	}
 
 	id, pipe, err := controls.NewPipe(r.pipes, req.AppID)
 	if err != nil {
-		xfer.ResponseError(err)
+		return xfer.ResponseError(err)
 	}
 	local, _ := pipe.Ends()
 	cw, err := r.client.StartExecNonBlocking(exec.ID, docker_client.StartExecOptions{

--- a/probe/docker/reporter.go
+++ b/probe/docker/reporter.go
@@ -20,14 +20,16 @@ const (
 type Reporter struct {
 	registry Registry
 	hostID   string
+	probeID  string
 	probe    *probe.Probe
 }
 
 // NewReporter makes a new Reporter
-func NewReporter(registry Registry, hostID string, probe *probe.Probe) *Reporter {
+func NewReporter(registry Registry, hostID string, probeID string, probe *probe.Probe) *Reporter {
 	reporter := &Reporter{
 		registry: registry,
 		hostID:   hostID,
+		probeID:  probeID,
 		probe:    probe,
 	}
 	registry.WatchContainerUpdates(reporter.ContainerUpdated)
@@ -103,9 +105,11 @@ func (r *Reporter) containerTopology(localAddrs []net.IP) report.Topology {
 		Icon:  "fa-terminal",
 	})
 
+	metadata := map[string]string{report.ControlProbeID: r.probeID}
+
 	r.registry.WalkContainers(func(c Container) {
 		nodeID := report.MakeContainerNodeID(c.ID())
-		result.AddNode(nodeID, c.GetNode(r.hostID, localAddrs))
+		result.AddNode(nodeID, c.GetNode(r.hostID, localAddrs).WithLatests(metadata))
 	})
 
 	return result

--- a/probe/docker/reporter_test.go
+++ b/probe/docker/reporter_test.go
@@ -50,8 +50,10 @@ var (
 )
 
 func TestReporter(t *testing.T) {
+	var controlProbeID = "a1b2c3d4"
+
 	containerImageNodeID := report.MakeContainerImageNodeID("baz")
-	rpt, err := docker.NewReporter(mockRegistryInstance, "host1", "probeID", nil).Report()
+	rpt, err := docker.NewReporter(mockRegistryInstance, "host1", controlProbeID, nil).Report()
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -65,9 +67,10 @@ func TestReporter(t *testing.T) {
 		}
 
 		for k, want := range map[string]string{
-			docker.ContainerID:   "ping",
-			docker.ContainerName: "pong",
-			docker.ImageID:       "baz",
+			docker.ContainerID:    "ping",
+			docker.ContainerName:  "pong",
+			docker.ImageID:        "baz",
+			report.ControlProbeID: controlProbeID,
 		} {
 			if have, ok := node.Latest.Lookup(k); !ok || have != want {
 				t.Errorf("Expected container %s latest %q: %q, got %q", containerNodeID, k, want, have)

--- a/probe/docker/reporter_test.go
+++ b/probe/docker/reporter_test.go
@@ -51,7 +51,7 @@ var (
 
 func TestReporter(t *testing.T) {
 	containerImageNodeID := report.MakeContainerImageNodeID("baz")
-	rpt, err := docker.NewReporter(mockRegistryInstance, "host1", nil).Report()
+	rpt, err := docker.NewReporter(mockRegistryInstance, "host1", "probeID", nil).Report()
 	if err != nil {
 		t.Fatal(err)
 	}

--- a/probe/host/tagger.go
+++ b/probe/host/tagger.go
@@ -9,15 +9,13 @@ import (
 // in every topology to an origin host node in the host topology.
 type Tagger struct {
 	hostNodeID string
-	probeID    string
 }
 
 // NewTagger tags each node with a foreign key linking it to its origin host
 // in the host topology.
-func NewTagger(hostID, probeID string) Tagger {
+func NewTagger(hostID string) Tagger {
 	return Tagger{
 		hostNodeID: report.MakeHostNodeID(hostID),
-		probeID:    probeID,
 	}
 }
 
@@ -27,11 +25,8 @@ func (Tagger) Name() string { return "Host" }
 // Tag implements Tagger.
 func (t Tagger) Tag(r report.Report) (report.Report, error) {
 	var (
-		metadata = map[string]string{
-			report.HostNodeID: t.hostNodeID,
-			report.ProbeID:    t.probeID,
-		}
-		parents = report.EmptySets.Add(report.Host, report.MakeStringSet(t.hostNodeID))
+		metadata = map[string]string{report.HostNodeID: t.hostNodeID}
+		parents  = report.EmptySets.Add(report.Host, report.MakeStringSet(t.hostNodeID))
 	)
 
 	// Explicity don't tag Endpoints and Addresses - These topologies include pseudo nodes,

--- a/probe/host/tagger_test.go
+++ b/probe/host/tagger_test.go
@@ -10,25 +10,19 @@ import (
 func TestTagger(t *testing.T) {
 	var (
 		hostID         = "foo"
-		probeID        = "a1b2c3d4"
 		endpointNodeID = report.MakeEndpointNodeID(hostID, "1.2.3.4", "56789") // hostID ignored
 		node           = report.MakeNodeWith(map[string]string{"foo": "bar"})
 	)
 
 	r := report.MakeReport()
 	r.Process.AddNode(endpointNodeID, node)
-	rpt, _ := host.NewTagger(hostID, probeID).Tag(r)
+	rpt, _ := host.NewTagger(hostID).Tag(r)
 	have := rpt.Process.Nodes[endpointNodeID].Copy()
 
 	// It should now have the host ID
 	wantHostID := report.MakeHostNodeID(hostID)
 	if hostID, ok := have.Latest.Lookup(report.HostNodeID); !ok || hostID != wantHostID {
 		t.Errorf("Expected %q got %q", wantHostID, report.MakeHostNodeID(hostID))
-	}
-
-	// It should now have the probe ID
-	if haveProbeID, ok := have.Latest.Lookup(report.ProbeID); !ok || haveProbeID != probeID {
-		t.Errorf("Expected %q got %q", probeID, haveProbeID)
 	}
 
 	// It should still have the other keys

--- a/prog/probe.go
+++ b/prog/probe.go
@@ -155,7 +155,7 @@ func probeMain() {
 		if registry, err := docker.NewRegistry(*dockerInterval, clients); err == nil {
 			defer registry.Stop()
 			p.AddTagger(docker.NewTagger(registry, processCache))
-			p.AddReporter(docker.NewReporter(registry, hostID, p))
+			p.AddReporter(docker.NewReporter(registry, hostID, probeID, p))
 		} else {
 			log.Errorf("Docker: failed to start registry: %v", err)
 		}

--- a/prog/probe.go
+++ b/prog/probe.go
@@ -146,7 +146,7 @@ func probeMain() {
 		host.NewReporter(hostID, hostName),
 		process.NewReporter(processCache, hostID, process.GetDeltaTotalJiffies),
 	)
-	p.AddTagger(probe.NewTopologyTagger(), host.NewTagger(hostID, probeID))
+	p.AddTagger(probe.NewTopologyTagger(), host.NewTagger(hostID))
 
 	if *dockerEnabled {
 		if err := report.AddLocalBridge(*dockerBridge); err != nil {

--- a/render/detailed/node.go
+++ b/render/detailed/node.go
@@ -101,7 +101,10 @@ func controlsFor(topology report.Topology, nodeID string) []ControlInstance {
 
 	for _, id := range node.Controls.Controls {
 		if control, ok := topology.Controls[id]; ok {
-			probeID, _ := node.Latest.Lookup(report.ProbeID)
+			probeID, ok := node.Latest.Lookup(report.ControlProbeID)
+			if !ok {
+				continue
+			}
 			result = append(result, ControlInstance{
 				ProbeID: probeID,
 				NodeID:  nodeID,

--- a/report/report.go
+++ b/report/report.go
@@ -209,4 +209,6 @@ const (
 	HostNodeID = "host_node_id"
 	// ProbeID is the random ID of the probe which generated the specific node.
 	ProbeID = "probe_id"
+	// ControlProbeID is the random ID of the probe which controls the specific node.
+	ControlProbeID = "control_probe_id"
 )

--- a/report/report.go
+++ b/report/report.go
@@ -207,8 +207,6 @@ const (
 	// a node in the host topology. That host node is the origin host, where
 	// the node was originally detected.
 	HostNodeID = "host_node_id"
-	// ProbeID is the random ID of the probe which generated the specific node.
-	ProbeID = "probe_id"
 	// ControlProbeID is the random ID of the probe which controls the specific node.
 	ControlProbeID = "control_probe_id"
 )

